cluster/local/provider.Tiltfile

```diff
@@ -0,0 +1,196 @@
+k8s_kind('ControllerConfig', image_json_path = '{.spec.image}')
+
+# Build a crossplane provider
+#
+# 1. build and apply CRDs
+# 2. build and deploy controller manager
+def build_provider(name, provider, settings):
+    _generate_crds(provider, settings)
+    _build_controller(name, provider, settings)
+
+# Generate and apply crossplane provider's CRDs
+def _generate_crds(provider, settings):
+    context = provider.get('context')
+
+    # Prefix each dependency with context so Tilt can watch the correct paths
+    # for changes. At the same time keep track of 'zz_generated.*' files to
+    # ignore (this is needed to prevent being stuck in an infinite loop.)
+    crd_deps = []
+    ignore_deps = []
+    for d in provider.get('crd_deps', []):
+        crd_deps.append(context + '/' + d)
+        ignore_deps.append(context + '/' + d + '/**/zz_generated.*')
+        
+    # Generate crossplane provider CRDs out of API definition
+    local_resource(
+        'gen-crds-' + provider['short_name'],
+        'cd ' + context + ' && make generate',
+        deps = crd_deps,
+        ignore = ignore_deps
+    )
+
+    # Find all the CRDs manifests files
+    crds_folder = context + '/' + provider.get('crds_folder', 'package/crds')
+    files = listdir(crds_folder)
+
+    # Extract name of the CRDs to show on Tilt UI
+    crds_name = []
+    for f in files:
+        crds_name.append(read_yaml(f)['metadata']['name'] + ':customresourcedefinition')
+
+    # Apply CRD manifests
+    k8s_yaml(files)
+    k8s_resource(
+        new_name = 'crds-' + provider['short_name'],
+        objects = crds_name,
+        resource_deps = settings.get('resource_deps')
+    )
+
+# Build crossplane provider binary and docker image. Binary is built for
+# linux-amd64 (to be only used inside a container.) The binary is stored
+# in crossplane cloned folder at _output/tilt/
+def _build_controller(name, provider, settings):
+    context = provider.get('context')
+    go_main = provider.get('go_main')
+    core_path = settings.get('core_path')
+
+    # Prefix each live reload dependency with context so Tilt
+    # can watch the correct paths for changes
+    cmd_deps = []
+    ignore_deps = []
+    for d in provider.get('cmd_deps', []):
+        cmd_deps.append(context + '/' + d)
+        ignore_deps.append(context + '/' + d + '/**/zz_generated.*')
+
+    # Build crossplane provider
+    build_output = '_output/tilt/' + name
+    local_resource(
+        'build-' + provider['short_name'],
+        'cd ' + context + ' && CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o ' + core_path + '/' + build_output + ' ' + go_main,
+        deps = cmd_deps,
+        ignore = ignore_deps
+    )
+
+    # Build crossplane provider docker image
+    dockerfile_contents = '\n'.join([
+        'FROM alpine:3.7',
+        'RUN apk --no-cache add ca-certificates bash',
+        'ADD ' + build_output + ' /usr/local/bin/crossplane-provider',
```

Comment:
Can we use the same `Dockerfile` we use for actually building the images?

Reply:
Specifically in this file (i.e. `cluster/local/provider.Tiltfile`), no. Because this `dockerfile_contents` is intended for building and live-reloading. To clarify a bit more, this file is being used in two different contexts:

- inside `crossplane/crossplane` when someone wants to work on Crossplane *and* a provider at the same time
- inside `org/provider-foo`, when someone is only working on a provider with the assumption of crossplane is running

In the first case there is no visibility of `Dockerfile` for provider at all to do anything with it.

But in the second case you could argue the `Dockerfile` for provider already exists in the context, but the main (missing) enhancement point would be `tilt_start.sh` and `tilt_restart.sh` to live reload the process inside the running pod without bouncing the pod itself.